### PR TITLE
Introducing Capability API

### PR DESCRIPTION
**Pull request by guzzilar**
> Note: This pull request is branched from #99 as I'm taking an ownership of the pull request.
> Originally was proposed by @jonrandy (as well as the pull request description below).

Adds initial functionality for using the Capability API. Quite basic right now, probably needs to be expanded upon

Example usage:
```php
<?php

require_once 'lib/Omise.php';

define('OMISE_PUBLIC_KEY', 'pkey_xxxxxxxxxxxxx');
define('OMISE_SECRET_KEY', 'skey_xxxxxxxxxxxxx');



$capabilities = OmiseCapabilities::retrieve();

print_r($capabilities);


$instalmentBackends = $capabilities->getBackends(
	$capabilities->makeBackendFilterType('installment'),
	$capabilities->makeBackendFilterCurrency('THB'),
	$capabilities->makeBackendFilterChargeAmount(749999)
);


foreach ($instalmentBackends as $backend) {
	print_r($backend);
}

```

### CODE REVIEW

File: lib/omise/OmiseCapabilities.php [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+     * @var array  of the filterable keys.
+     */
+    const FILTERS = array(
+        'backend' => array('currency', 'type', 'chargeAmount')
+    );
+
+    protected function __construct($publickey = null, $secretkey = null)
+    {
+        parent::__construct($publickey, $secretkey);
+        $this->setupFilterShortcuts();
```
Comment (jonrandy) [inline lib/omise/OmiseCapabilities.php]: Again, I have a really big problem with not allowing underscore prefixes to name `protected` and `private` members. Without these, the code is less readable. I will continue to use them

File: lib/omise/OmiseCapabilities.php [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+    /**
+     * Retrieves array of payment backends. Optionally pass in as many filter functions as you want
+     * (muliple arguments, or a single array)
+     *
+     * @param [func1,fun2,...] OR func1, func2,...
+     *
+     * @return array
+     */
+    public function getBackends()
```
Comment (jonrandy) [inline lib/omise/OmiseCapabilities.php]: Why make this function so verbose? Was concise before

File: lib/omise/OmiseCapabilities.php [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+                return (object)reset($backend);
+            },
+            $this['payment_backends']
+        );
+
+        // check for filters
+        if ($filters = func_get_args()) {
+            $filter   = self::combineFilters(self::argsToVariadic($filters));
+            $backends = array_filter($backends, $filter);
+            sort($backends);
```
Comment (jonrandy) [inline lib/omise/OmiseCapabilities.php]: Why sort? Pointless?

File: lib/omise/OmiseCapabilities.php [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+    /**
+     * Makes a filter function to check supported currency for backend.
+     *
+     * @param  string $currency
+     *
+     * @return function
+     */
+    public function makeBackendFilterCurrency($currency)
+    {
+        return function ($backend) use ($currency) {
```
Comment (jonrandy) [inline lib/omise/OmiseCapabilities.php]: Again, why split over many lines? It's like code diarrhoea

File: lib/omise/OmiseCapabilities.php [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+     * Makes a filter function to check type of backend.
+     *
+     * @param  string $type
+     *
+     * @return function
+     */
+    public function makeBackendFilterType($type)
+    {
+        return function ($backend) use ($type) {
+            return $backend->type == $type;
```
Comment (jonrandy) [inline lib/omise/OmiseCapabilities.php]: Code diarrhoea

File: lib/omise/OmiseCapabilities.php [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+        };
+    }
+
+    /**
+     * Sets up 'shortcuts' to filters so they may be used thus:
+     *    $capabilities->backendFilter['currency']('THB')
+     * As well as the original:
+     *    $capabilities->makeBackendFilterCurrency('THB')
+     */
+    protected function setupFilterShortcuts()
```
Comment (jonrandy) [inline lib/omise/OmiseCapabilities.php]: Better with _ prefix. Also, why move this function down here, better to leave it near where it was called (in the constructor)? Makes editing the file easier - less jumping around

File: lib/omise/OmiseCapabilities.php [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+    {
+        foreach (self::FILTERS as $filterSubject => $availableFilters) {
+            $filterArrayName = $filterSubject . 'Filter';
+            $this->$filterArrayName = array();
+            $tempArr = &$this->$filterArrayName;
+            foreach ($availableFilters as $type) {
+                $funcName = 'make' . ucfirst($filterSubject) . 'Filter' . $type;
+                $tempArr[$type] = function () use ($funcName) {
```
Comment (jonrandy) [inline lib/omise/OmiseCapabilities.php]: Why spread a single operation function across so many lines?

File: lib/omise/OmiseCapabilities.php [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+    }
+
+    /**
+     * Combines boolean filters.
+     *
+     * @param  [functions] $filters
+     *
+     * @return function
+     */
+    private static function combineFilters($filters)
```
Comment (jonrandy) [inline lib/omise/OmiseCapabilities.php]: Naming _

File: lib/omise/OmiseCapabilities.php [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+    {
+        return function ($value) use ($filters) {
+            foreach ($filters as $filter) {
```
Comment (jonrandy) [inline lib/omise/OmiseCapabilities.php]: Code diarrhoea. The code equivalent of div-itis in HTML

File: lib/omise/OmiseCapabilities.php [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+    }
+
+    /**
+     * Converts args to variadic fashion, rather than as a single array
+     *
+     * @param  [functions] $filters
+     *
+     * @return function
+     */
+    private static function argsToVariadic($argArray)
```
Comment (jonrandy) [inline lib/omise/OmiseCapabilities.php]: _

File: lib/omise/OmiseCapabilities.php (hunk 1 above)
Comment (jacstn) [reply lib/omise/OmiseCapabilities.php]: @jonrandy https://www.php-fig.org/psr/psr-2/#43-methods

File: lib/omise/OmiseCapabilities.php (hunk 4 above)
Comment (guzzilar) [reply lib/omise/OmiseCapabilities.php]: All of code diarrhoea stuff you commented were changed for the sake of Code Sniffer to make it comply with the code standard.

File: lib/omise/OmiseCapabilities.php (hunk 6 above)
Comment (guzzilar) [reply lib/omise/OmiseCapabilities.php]: Just moved those protected private stuff down to bottom and moved public methods up above.

File: lib/omise/OmiseCapabilities.php (hunk 1 above)
Comment (jonrandy) [reply lib/omise/OmiseCapabilities.php]: I know, but I have a really big problem with it and I don't think we should do it

File: lib/omise/OmiseCapabilities.php (hunk 7 above)
Comment (guzzilar) [reply lib/omise/OmiseCapabilities.php]: PHP Code Sniffer reason (same as the one I commented above).

File: lib/omise/OmiseCapabilities.php (hunk 9 above)
Comment (guzzilar) [reply lib/omise/OmiseCapabilities.php]: PHP Code Sniffer reason (same as the one I commented above).

File: lib/omise/OmiseCapabilities.php (hunk 1 above)
Comment (guzzilar) [reply lib/omise/OmiseCapabilities.php]: Again, this topic depends on personal taste. So I picked the one without `_` to make code consistency with the rest. And also to comply with PSR-2 & PHP Code Sniffer standard.

Comment (jonrandy) [reply lib/omise/OmiseCapabilities.php]: Code Sniffer can use any set of rules. We could inherit from PSR-2 and switch off rules we don't like

File: lib/omise/OmiseCapabilities.php (hunk 6 above)
Comment (jonrandy) [reply lib/omise/OmiseCapabilities.php]: Disagree with this too. Better to group related methods together

File: lib/omise/OmiseCapabilities.php (hunk 7 above)
Comment (jonrandy) [reply lib/omise/OmiseCapabilities.php]: Drop this rule from our CodeSniffer ruleset?

File: lib/omise/OmiseCapabilities.php (hunk 3 above)
Comment (jonrandy) [reply lib/omise/OmiseCapabilities.php]: It may actually change the order in which they were in the API response? Regardless, it seems unnecessary

Comment (guzzilar) [reply lib/omise/OmiseCapabilities.php]: Have no special reason for now than want to make the index sorted.
I'll remove this line then.

**Pull request merged**